src/engine/animation/animation.game.ts

```diff
@@ -0,0 +1,303 @@
+import { PhysicsModule, RigidBody } from "../physics/physics.game";
+
+export interface IAnimationComponent {
+  mixer: AnimationMixer;
+  clips: AnimationClip[];
+  actions: Map<String, AnimationAction>;
+}
+
+export enum AnimationClipType {
+  Fall2 = "Fall2",
+  Fall1 = "Fall1",
+  TurnLeft = "TurnLeft",
+  TurnRight = "TurnRight",
+  Idle = "Idle",
+  StrafeLeft = "StrafeLeft",
+  StrafeRight = "StrafeRight",
+  Walk = "Walk",
+  WalkBack = "WalkBack",
+  StrafeLeftRun = "StrafeLeftRun",
+  StrafeRightRun = "StrafeRightRun",
+  Run = "Run",
+  RunBack = "RunBack",
+}
+
+export const AnimationComponent = new Map<number, IAnimationComponent>();
+export const BoneComponent = new Map<number, Bone>();
+
+const animationQuery = defineQuery([AnimationComponent]);
+const enterAnimationQuery = enterQuery(animationQuery);
+// const exitAnimationQuery = exitQuery(animationQuery);
+const boneQuery = defineQuery([BoneComponent]);
+
+/*
+
+notes on calculating forward/up/right:
+
+  forward.x =  cos(pitch) * sin(yaw);
+  forward.y = -sin(pitch);
+  forward.z =  cos(pitch) * cos(yaw);
+
+  right.x =  cos(yaw);
+  right.y =  0;
+  right.z = -sin(yaw);
+
+  up = cross(forward, right);
+
+  equivalent:
+  up.x = sin(pitch) * sin(yaw);
+  up.y = cos(pitch);
+  up.z = sin(pitch) * cos(yaw);
+
+*/
+
+export function AnimationSystem(ctx: GameState) {
+  initializeAnimations(ctx);
+  processAnimations(ctx);
+  syncBones(ctx);
+}
+
+const _vel = vec3.create();
+const _forward = vec3.create();
+const _right = vec3.create();
+
+// TODO: reuse isGrounded for char controller
+const interactionGroup = 0xf00f_f00f;
+const colliderShape = new Capsule(0.5, 0.5);
+const shapeTranslationOffset = new Vector3(0, 0, 0);
+const shapeRotationOffset = new Quaternion(0, 0, 0, 0);
+const shapeCastPosition = new Vector3();
+const shapeCastRotation = new Quaternion();
+const _obj = new Object3D();
+
+const isGrounded = (ctx: GameState, physicsWorld: RAPIER.World, body: RAPIER.RigidBody) => {
+  shapeCastPosition.copy(body.translation() as Vector3).add(shapeTranslationOffset);
+  shapeCastRotation.copy(_obj.quaternion).multiply(shapeRotationOffset);
+
+  const shapeCastResult = physicsWorld.castShape(
+    shapeCastPosition,
+    shapeCastRotation,
+    physicsWorld.gravity,
+    colliderShape,
+    ctx.dt,
+    interactionGroup
+  );
+
+  const isGrounded = !!shapeCastResult;
+
+  return isGrounded;
+};
+
+const idleThreshold = 0.5;
+const walkThreshold = 10;
+
+const fadeInAmount = 8;
+const fadeOutAmount = fadeInAmount / 2;
+
+const lastYrot = new Float32Array(maxEntities);
+
+function initializeAnimations(ctx: GameState) {
+  const entered = enterAnimationQuery(ctx.world);
+  for (let i = 0; i < entered.length; i++) {
+    const eid = entered[i];
+    const animation = AnimationComponent.get(eid);
+
+    if (animation) {
+      animation.actions = animation.clips.reduce((map, clip) => {
+        const action = animation.mixer.clipAction(clip).play();
```

Comment:
Should we still be playing animations automatically?

Reply:
yes, but they don't need to be `enabled` all the time. i added appropriate enables/disables, should save the mixer a few cycles